geco/mips/facility_location/orlib.py

```diff
@@ -0,0 +1,134 @@
+import pyscipopt as scip
+import geco.mips.loading.orlib as orlib
+from geco.mips.facility_location.generic import capacitated_warehouse_location
+
+
+def _cap_numeric_reader(file):
+    """
+    Reads cap(NUMBER) Capacitated Warehouse Location instances mentioned in [1].
+
+    Parameters
+    ----------
+    file: file-like object
+
+    Returns
+    -------
+    model: scip.Model
+        A pyscipopt model of the loaded instance
+
+    References
+    ----------
+    ..[1] http://people.brunel.ac.uk/~mastjjb/jeb/orlib/capinfo.html
+    """
+    num_of_warehouses, num_of_customers = orlib.read_numbers(file.readline())
+    capacities = []
+    fixed_costs = []
+    demands = []
+    allocation_cost_per_warehouse = {}
+
+    for _ in range(num_of_warehouses):
+        capacity, fixed_cost = orlib.read_numbers(file.readline())
+        capacities.append(capacity)
+        fixed_costs.append(fixed_cost)
+
+    for i in range(num_of_customers):
+        demand = orlib.read_number(file.readline())
+        allocation_costs = orlib.read_multiline_numbers(file, num_of_warehouses)
+        demands.append(demand)
+        for j, cost in enumerate(allocation_costs):
+            allocation_cost_per_warehouse[i, j] = cost
+
+    return capacitated_warehouse_location(
+        n_customers=num_of_customers,
+        n_facilities=num_of_warehouses,
+        transportation_cost=allocation_cost_per_warehouse,
+        demands=demands,
+        capacities=capacities,
+        fixed_costs=fixed_costs,
+    )
+
+
+def _cap_alpha_reader(file, capacity):
+    """
+    Reads cap(LETTER) Capacitated Warehouse Location instances mentioned in [1].
+
+    Parameters
+    ----------
+    file: file-like object
+
+    Returns
+    -------
+    model: scip.Model
+        A pyscipopt model of the loaded instance
+
+    References
+    ----------
+    ..[1] http://people.brunel.ac.uk/~mastjjb/jeb/orlib/capinfo.html
+    """
+    num_of_warehouses, num_of_customers = orlib.read_numbers(file.readline())
+    capacities = []
+    fixed_costs = []
+    demands = []
+    allocation_cost_per_warehouse = {}
+
+    for _ in range(num_of_warehouses):
+        capacity, fixed_cost = capacity, float(file.readline().strip().split(b" ")[-1])
+        capacities.append(capacity)
+        fixed_costs.append(fixed_cost)
+
+    for i in range(num_of_customers):
+        demand = orlib.read_number(file.readline())
+        allocation_costs = orlib.read_multiline_numbers(file, num_of_warehouses)
+        demands.append(demand)
+        for j, cost in enumerate(allocation_costs):
+            allocation_cost_per_warehouse[i, j] = cost
+
+    return capacitated_warehouse_location(
```

Comment:
Same as above (I think we should return all parameters and leave it to the user which formulation to pass them to).